File: packages/ts/src/named-positions.ts
// Canonical OCF named positions come from the spec catalog (language-neutral data
// in shared/named-positions.json), NOT from the JSON schema — the schema's
// coordinate_named.named is a free-form string with no enum. The Python mirror
// loads the same JSON file. Document-declared custom positions (named_positions.custom)
// are merged in at runtime.
import catalog from "../../../shared/named-positions.json" with { type: "json" };

export const CANONICAL_NAMED: Set<string> = new Set(
  (catalog as { positions: string[] }).positions,
);

export function knownNamed(doc: Record<string, unknown>): Set<string> {
  const set = new Set(CANONICAL_NAMED);
  const np = (doc.named_positions ?? {}) as Record<string, unknown>;
  const custom = (np.custom ?? {}) as Record<string, unknown>;
  for (const name of Object.keys(custom)) set.add(name);
  return set;
}

File: packages/py/ocf_validator/named_positions.py
import json
from pathlib import Path
from typing import Any

# Canonical OCF named positions come from the spec catalog (language-neutral
# data in shared/named-positions.json), NOT from the JSON schema. Mirrors the
# TS implementation, which loads the same JSON file.
_CATALOG = json.loads(
    (Path(__file__).resolve().parents[3] / "shared" / "named-positions.json").read_text()
)

CANONICAL_NAMED: set[str] = set(_CATALOG["positions"])


def known_named(doc: dict[str, Any]) -> set[str]:
    result = set(CANONICAL_NAMED)
    np = doc.get("named_positions") or {}
    custom = np.get("custom") if isinstance(np, dict) else None
    if isinstance(custom, dict):
        for name in custom.keys():
            result.add(name)
    return result
